### ISSUE TITLE
Question on Emby

### ISSUE THREAD

**Issue by feerlessleadr** [jellyfin/mopidy-jellyfin#17]
I realize that this may receive a negative reaction, but I am looking to set up mopidy (as part of a multiroom audio solution), and I'd love to be able to use Emby as the backend/source for mopidy. 

I currently use Emby for my media consumption needs (no ill will or anything with Jellyfin, just a personal choice). 

I know this is specific for Jellyfin, but I was wondering if you foresee any problem using this plugin with emby, since I think I read on reddit that the base of this plugin started with the (now deprecated) emby version of this plugin.

If not, is there any chance that you'd be willing to support emby in addition to Jellyfin?

Thanks for your consideration.

**Comment by mcarlton00** [comment]
Oddly enough, that's a future project of mine that I haven't gotten started on, combining this with [Snapcast](https://github.com/badaix/snapcast).  If you do go down that path, I'd be interested in any feedback you find with the process.

You're correct in that this started as a fork of [mopidy-emby](https://github.com/xsteadfastx/mopidy-emby).   It was never announced as depreciated (as far as I know), but it doesn't seem to be actively maintained anymore.

At this point, I have no plans to purposely break Emby compatibility.  In it's current state (version 0.6.x) it _should_ continue working without any problems.  That being said, sooner or later the API between Jellyfin and Emby is likely going to diverge, and obviously this project will follow where Jellyfin's implementation leads.  If both can be supported without a bunch of extra overhead, then I have no objections to continuing support.  But I don't have an Emby server to test against anymore, so I'd have to rely on user reports.

TLDR: "It works now, but I can't make any promises for long term support."

**Comment by feerlessleadr** [comment]
Appreciate the feedback. I'm planning on using snapcast as well, I'll report back on any issues/quirks that I come across. 

Also, loud and clear on "it works now, but may not in the future". 

I'll report any bugs on the emby side, and if they can be fixed, great, if not, then I'll survive.

**Comment by mcarlton00** [comment]
Or I'll use the allure of this client to drag you over to the Jellyfin side.  Whatever works, I'm not picky.

Good luck with the house audio, I'm looking forward to hearing about it down the road

**Comment by feerlessleadr** [comment]
so I am a total noob at linux, and since I could never get emby to work right, I figured that I would install jellyfin to test out for my music purposes. 

The thing is, I can't for the life of me get Jellyfin to work either it seems. I have mopidy working correctly with local files, but after I installed and configured this, nothing shows up in either MALP or my web frontend. 

I am running mopidy as a service, and I have configured jellyfin in `/etc/mopidy/mopidy.conf` as follows:
```
[core]
cache_dir = /var/cache/mopidy
config_dir = /etc/mopidy
data_dir = /var/lib/mopidy
max_tracklist_length = 10000
restore_state = false

[logging]
color = true
console_format = %(levelname)-8s %(message)s
debug_format = %(levelname)-8s %(asctime)s [%(process)d:%(threadName)s] %(name)s\n  %(message)s
debug_file = /var/log/mopidy/mopidy-debug.log
config_file = /etc/mopidy/logging.conf

[audio]
mixer = software
mixer_volume =
#output = autoaudiosink
output = audioresample ! audio/x-raw,rate=48000,channels=2,format=S16LE ! audioconvert ! wavenc ! filesink location=/tmp/snapfifo
buffer_time =

[proxy]
scheme =
hostname =
port =
username =
password =

[mpd]
enabled = true
hostname = ::
port = 6600
password =
max_connections = 20
connection_timeout = 60
zeroconf = Mopidy MPD server on $hostname
command_blacklist =
  listall
  listallinfo
default_playlist_scheme = m3u

[http]
enabled = true
hostname = ::
port = 6680
static_dir =
zeroconf = Mopidy HTTP server on $hostname

[stream]
enabled = true
protocols =
  http
  https
  mms
  rtmp
  rtmps
  rtsp
metadata_blacklist =
timeout = 5000

[m3u]
enabled = true
base_dir = $XDG_MUSIC_DIR
default_encoding = latin-1
default_extension = .m3u8
playlists_dir = /media/v/Music/Playlists

[softwaremixer]
enabled = true

[file]
enabled = true
media_dirs =
  $XDG_MUSIC_DIR|Music
  ~/|Home
excluded_file_extensions =
  .jpg
  .jpeg
show_dotfiles = false
follow_symlinks = false
metadata_timeout = 1000

[local]
enabled = false
library = json
#media_dir = /var/lib/mopidy/media
media_dir = /media/v/Music
scan_timeout = 2000
scan_flush_threshold = 100
scan_follow_symlinks = false
excluded_file_extensions =
  .directory
  .html
  .jpeg
  .jpg
  .log
  .nfo
  .png
  .txt

[jellyfin]
enabled = true
hostname = 192.168.1.16
port = 8097
username = Kevin
password = XXXXXX (blanked out for purposes of posting)
libraries = 
```
However, when I run `sudo mopidyctl config`, the jellyfin section does not show up in the output, so it is not being read for some reason. I have restarted the service multiple times to no avail. 

Here is the output for `sudo service mopidy status` which I believe confirms that the jellyfin backend is not being picked up by the service
```
kevin@linuxserver:~$ sudo service mopidy status
● mopidy.service - Mopidy music server
   Loaded: loaded (/lib/systemd/system/mopidy.service; enabled; vendor preset: enabled)
   Active: active (running) since Sat 2020-01-04 02:00:37 EST; 6min ago
  Process: 1453 ExecStartPre=/bin/chown mopidy:audio /var/cache/mopidy (code=exited, status=0/SUCCESS)
  Process: 1433 ExecStartPre=/bin/mkdir -p /var/cache/mopidy (code=exited, status=0/SUCCESS)
 Main PID: 1491 (mopidy)
    Tasks: 12 (limit: 4279)
   CGroup: /system.slice/mopidy.service
           └─1491 /usr/bin/python /usr/bin/mopidy --config /usr/share/mopidy/conf.d:/etc/mopidy/mopidy.conf

Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Starting Mopidy mixer: SoftwareMixer
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Starting Mopidy audio
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Starting Mopidy backends: StreamBackend, M3UBackend, FileBackend
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Audio output set to "audioresample ! audio/x-raw,rate=48000,channels=2,format=S16LE ! audioconvert ! wavenc ! filesink location=/tmp/snapfi
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Starting Mopidy core
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Starting Mopidy frontends: IrisFrontend, MpdFrontend, HttpFrontend
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     MPD server running at [::]:6600
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Starting Iris 3.43.0
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     Starting GLib mainloop
Jan 04 02:00:41 linuxserver mopidy[1491]: INFO     HTTP server running at [::]:6680
```

When I run `sudo mopidy config` (for the non service config), just for kicks, even though the conf file for the non service option is basically blank (since I run mopidy as a service), the Jellyfin section DOES show up:
```
[jellyfin]
enabled = false  ; Extension disabled due to config errors.
username =   ; Must be set.
user_id =
password =   ; Must be set.
hostname =   ; Must be set.
libraries =
port =   ; Must be set.
client_cert =
client_key =
```
Any idea what I'm doing wrong?

**Comment by mcarlton00** [comment]
A quick test to show if it's connecting properly is to browse through "Files -> Jellyfin".  It should list all of your "Music" and "Book" type libraries there and let you navigate the file structure.  If you can browse through Files, but nothing is appearing in the library, it means that you'll need to adjust how the libraries are configured.

No concrete ideas off the top of my head, but we should be able to track it down.  I would check the log file first and see if anything is appearing there with reasons for the failure.  Alternatively, you can stop the service and run `sudo mopidy --config /etc/mopidy/mopidy.conf` so it should post output to the screen (you may have to disable the logging section for that to work, I'm not 100% sure).

The next option is how your Jellyfin server is setup.  As mentioned above, the plugin only handles Music and Book libraries, so if you've added it as a Mixed Content type or something else it likely won't work (I haven't tested).  Also, if you don't fill in the "libraries =" line in the config file, it only looks for a library named "Music".  If your attached library is named otherwise you'll need to configure that option.

Final point for right now is that both username and libraries are case sensitive.  ie: 'Kevin' != 'kevin', 'Music' != 'music', etc.

**Comment by feerlessleadr** [comment]
So when I browse to files, I don't have Jellyfin listed, it just has local, local-images, media, & playlists. I'm 99% sure that for some reason the jellyfin module isn't being loaded by the service for some reason. 

This is what the log shows from `/var/log/mopidy/mopidy.log`
```
2020-01-04 12:05:36,409 INFO [22537:MainThread] mopidy.__main__: Starting Mopidy 2.1.0
2020-01-04 12:05:36,756 INFO [22537:MainThread] mopidy.config: Loading config from builtin defaults
2020-01-04 12:05:36,758 INFO [22537:MainThread] mopidy.config: Loading config from /etc/mopidy/mopidy.conf
2020-01-04 12:05:36,760 INFO [22537:MainThread] mopidy.config: Loading config from command line options
2020-01-04 12:05:36,802 INFO [22537:MainThread] mopidy.__main__: Enabled extensions: iris, mpd, http, file, stream, m3u, softwaremixer, moped, local-images
2020-01-04 12:05:36,802 INFO [22537:MainThread] mopidy.__main__: Disabled extensions: local
2020-01-04 12:05:36,850 INFO [22537:MainThread] mopidy.commands: Starting Mopidy mixer: SoftwareMixer
2020-01-04 12:05:36,852 INFO [22537:MainThread] mopidy.commands: Starting Mopidy audio
2020-01-04 12:05:36,853 INFO [22537:MainThread] mopidy.commands: Starting Mopidy backends: StreamBackend, M3UBackend, FileBackend
2020-01-04 12:05:36,863 INFO [22537:MainThread] mopidy.commands: Starting Mopidy core
2020-01-04 12:05:36,883 INFO [22537:Audio-2] mopidy.audio.actor: Audio output set to "audioresample ! audio/x-raw,rate=48000,channels=2,format=S16LE ! audioconvert ! wavenc ! filesink location=/tmp/snapfifo"
2020-01-04 12:05:37,128 INFO [22537:MainThread] mopidy.commands: Starting Mopidy frontends: IrisFrontend, MpdFrontend, HttpFrontend
2020-01-04 12:05:37,130 INFO [22537:MainThread] mopidy.mpd.actor: MPD server running at [::]:6600
2020-01-04 12:05:37,131 INFO [22537:IrisFrontend-7] mopidy_iris.core: Starting Iris 3.43.0
2020-01-04 12:05:37,133 INFO [22537:MainThread] mopidy.commands: Starting GLib mainloop
2020-01-04 12:05:37,133 INFO [22537:HttpFrontend-10] mopidy.http.actor: HTTP server running at [::]:6680
2020-01-04 12:05:37,488 INFO [22537:HttpServer] tornado.access: 101 GET /iris/ws/ (192.168.1.16) 0.96ms
```
It looks like the jellyfin extension isn't being picked up at all either under enabled or disabled extensions, even though it is installed:
```
kevin@linuxserver:~$ pip show mopidy-jellyfin
Name: Mopidy-Jellyfin
Version: 0.6.0
Summary: Mopidy extension for playing music from jellyfin
Home-page: https://github.com/mcarlton00/mopidy-jellyfin
Author: Matt Carlton
Author-email: mcarlton00@gmail.com
License: Apache License, Version 2.0
Location: /home/kevin/.local/lib/python2.7/site-packages
Requires: unidecode, Mopidy, setuptools, Pykka, requests
```

When I run mopidy manually, the jellyfin module is picked up, and I see a Jellyfin section on my web interface (not all of my music is showing up, but that is a problem for after I can get the service working). I had to add in the user_id into the mopidy config for the manual run to work, but that had no affect on the service. 
```
kevin@linuxserver:~$ sudo mopidy --config /etc/mopidy/mopidy.conf
/usr/lib/python2.7/dist-packages/mopidy/ext.py:202: PkgResourcesDeprecationWarning: Parameters to load are deprecated.  Call .resolve and .require separately.
  extension_class = entry_point.load(require=False)
INFO     Starting Mopidy 2.1.0
INFO     Loading config from builtin defaults
INFO     Loading config from /etc/mopidy/mopidy.conf
INFO     Loading config from command line options
INFO     Enabled extensions: iris, mpd, http, moped, stream, m3u, softwaremixer, file, local-images, jellyfin
INFO     Disabled extensions: local
INFO     Starting Mopidy mixer: SoftwareMixer
INFO     Starting Mopidy audio
INFO     Starting Mopidy backends: StreamBackend, M3UBackend, FileBackend, JellyfinBackend
INFO     Audio output set to "audioresample ! audio/x-raw,rate=48000,channels=2,format=S16LE ! audioconvert ! wavenc ! filesink location=/tmp/snapfifo"
INFO     Starting Mopidy core
INFO     Starting Mopidy frontends: MpdFrontend, HttpFrontend, IrisFrontend
INFO     MPD server running at [::]:6600
INFO     HTTP server running at [::]:6680
INFO     Starting Iris 3.43.0
INFO     Starting GLib mainloop
```

**Comment by mcarlton00** [comment]
I've already moved to mopidy 3 on my speaker and dev systems, but using `mopidyctl` keeps throwing errors about missing pieces.  I'm not 100% sure what it's use case is, as I've always just used `mopidy` instead.  Is your service file specifically calling `mopidyctl`?  On my Pi the file is at `/etc/init.d/mopidy` and seems to only get called if you do a `service mopidy run` instead of `service mopidy start`.

Another thought: Did you `sudo pip install mopidy-jellyfin`, or `pip install --user mopidy-jellyfin`.  If you installed with the user flag, it's possible that `mopidy` and `mopidyctl` are being executed as different users.  That would explain one one of them can see the plugin but the other can't.

As for things not showing up all the way on the web frontend, I also had some problems with mopidy-iris and getting artists/albums to populate.  I commented about it [here](https://github.com/jaedb/Iris/issues/415#issuecomment-552065900) and just haven't gone back to figure out what option is needed to make it work.

**Comment by feerlessleadr** [comment]
Thanks - I used mopidyctl based on this guide/doc from mopidy about running mopidy as a service:

https://docs.mopidy.com/en/latest/running/service/

I'm still running 18.04, so I won't be able to update to mopidy 3 unless I move to a more current version of ubuntu. Do you think that could be the problem (i'm running a much older version of mopidy vs current release)?

**Comment by mcarlton00** [comment]
Hmm, I wonder if that page is accurate.  I didn't use it on my Pi or Ubuntu machines, but just `apt install mopidy` put a service file there for me automatically.  The file from my Ubuntu 18.04 machine is [here](https://pastebin.com/XWbnM94W)

If you'd like to move to mopidy 3 now, you can if you use their PPA: https://docs.mopidy.com/en/latest/installation/debian/#install-from-apt-mopidy-com.  This is the big transition to python 3, so there may be some other odd bugs lingering about (likely the cause of my `mopidyctl` weirdness I mentioned).  I haven't found any breaking bugs yet though.  Versions 0.5.5 of mopidy-jellyfin and later are mopidy3 compatible

**Comment by feerlessleadr** [comment]
Thanks - I have a feeling that this is all related to me being on 18.04 vs a later version of ubuntu (I do need to use mopidyctl otherwise my configs don't update). Let me update to a later version of ubuntu and try again, see if the problem persists.

**Comment by mcarlton00** [comment]
It's possible that you're just over complicating something in the process.  For example, my `/etc/mopidy/mopidy.conf` on a raspberry pi:
```
[mpd]
enabled = true
hostname = 0.0.0.0
connection_timeout = 60

[http]
hostname = 0.0.0.0
allowed_origins = pi.example.com

[jellyfin]
hostname = https://jellyfin.example.com
port = 443
username = Matt
password = supersecretpassword

[audio]
output = alsasink
```

I'm not sure what you mean by "configs don't update".  This config file hasn't been edited in several months, so I'm not sure what updates you're referring to.

**Comment by feerlessleadr** [comment]
Something must be wrong with my mopidy install, since I basically copied your config to mine, restarted the service, and jellyfin still isn't being picked up by mopidy (no errors, nothing, like it isn't even installed). 

[I have a topic on the Mopidy discourse forum](https://discourse.mopidy.com/t/mopidy-extension-not-loading-when-run-as-service/3684) to see if there is something on the mopidy side that I'm missing.

**Comment by feerlessleadr** [comment]
Ok, I was able to get everything sorted out. It turns out that when I installed Mopiby-Jellyfin, since I didn't install using sudo, it only installed for the current user, rather than all users. 

Once I re-installed using `sudo pip install Mopidy-Jellyfin` everything works correctly now. 

[See here for the discourse thread from one of the Mopidy devs](https://discourse.mopidy.com/t/mopidy-extension-not-loading-when-run-as-service/3684/3). 

I still have an issue where none of my players (M.A.L.P, Iris, & moped) have my music/playlists show up under library (and only a limited number of folders show up under files), but I'll create a separate issue for that for tracking purposes.
